src/UserSpaceInstrumentation/Attach.cpp

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "UserSpaceInstrumentation/Attach.h"
+
+#include <absl/strings/str_format.h>
+#include <errno.h>
+#include <stdint.h>
+#include <sys/ptrace.h>
+#include <sys/wait.h>
+
+#include <chrono>
+#include <memory>
+#include <set>
+#include <string>
+#include <thread>
+#include <vector>
+
+#include "OrbitBase/GetProcessIds.h"
+#include "OrbitBase/Logging.h"
+
+namespace orbit_user_space_instrumentation {
+
+using orbit_base::GetTidsOfProcess;
+
+namespace {
+
+// Attaches to the thread using ptrace.
+// Returns true if the thread was halted and false if the thread did not exist anymore.
+[[nodiscard]] ErrorMessageOr<bool> AttachAndStopThread(pid_t tid) {
+  if (ptrace(PTRACE_ATTACH, tid, nullptr, nullptr) == -1) {
+    // If tid has ended already we get ESRCH; if the thread was in 'exit state' we get EPERM.
+    // There are a bunch of other (non-relevant) cases. I haven't found documentation on this but it
+    // can be looked up in the function 'ptrace_attach' in 'ptrace.c' in the kernel sources.
+    if (errno == ESRCH || errno == EPERM) {
+      return false;
+    } else {
+      return ErrorMessage(
+          absl::StrFormat("PTRACE_ATTACH failed for \"%d\": \"%s\"", tid, strerror(errno)));
+    }
+  }
+  // Wait for the traced thread to stop. Timeout after one second.
+  int32_t timeout_ms = 1000;
+  while (true) {
+    int stat_val = 0;
+    const int result_waitpid = waitpid(tid, &stat_val, WNOHANG);
+    if (result_waitpid == -1) {
+      return ErrorMessage(absl::StrFormat(
+          "Wait for thread to get traced failed for tid \"%d\": \"%s\"", tid, strerror(errno)));
+    }
+    if (result_waitpid > 0) {
+      // Occasionally the thread is active during PTRACE_ATTACH but terminates before it gets
+      // descheduled. So waitpid returns on exit of the thread instead of the expected stop.
+      if (WIFEXITED(stat_val)) {
+        return false;
+      } else if (WIFSTOPPED(stat_val)) {
+        return true;
+      } else {
+        return ErrorMessage(absl::StrFormat(
+            "Wait for thread to get traced yielded unexpected result for tid \"%d\": \"%d\"", tid,
+            stat_val));
+      }
+    }
+    std::this_thread::sleep_for(std::chrono::milliseconds(1));
+    if (--timeout_ms == 0) {
+      ptrace(PTRACE_DETACH, tid, nullptr, nullptr);
+      return ErrorMessage(
+          absl::StrFormat("Waiting for the traced thread \"%d\" to stop timed out.", tid));
+    }
+  }
+  return true;
```

Comment:
this line seems to be unreachable....

Reply:
We have UNREACHABLE() in `OrbitBase/Logging.h`, it shouldn't cause warnings.